service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/BlobCopier.java

```diff
@@ -0,0 +1,109 @@
+package bio.terra.workspace.service.resource.controlled.cloud.azure;
+
+import static java.util.stream.Collectors.groupingBy;
+
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.resource.controlled.cloud.azure.storage.ControlledAzureStorageResource;
+import bio.terra.workspace.service.resource.controlled.cloud.azure.storageContainer.ControlledAzureStorageContainerResource;
+import com.azure.core.util.polling.PollResponse;
+import com.azure.core.util.polling.SyncPoller;
+import com.azure.storage.blob.BlobContainerClient;
+import com.azure.storage.blob.models.BlobCopyInfo;
+import com.azure.storage.blob.models.BlobItem;
+import java.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BlobCopier {
+  private final Logger logger = LoggerFactory.getLogger(BlobCopier.class);
+  private final AzureStorageAccessService storageAccessService;
+  private final AuthenticatedUserRequest userRequest;
+
+  private static final Duration MAX_POLL_TIMEOUT = Duration.ofMinutes(10);
+
+  public BlobCopier(
+      AzureStorageAccessService storageAccessService, AuthenticatedUserRequest userRequest) {
+    this.storageAccessService = storageAccessService;
+    this.userRequest = userRequest;
+  }
+
+  /**
+   * Attempts to copy all blobs from the source container to the destination container. The calling
+   * principal must have READ-level access on the storage container resource.
+   *
+   * @param sourceStorageAccount Azure storage account for the source container
+   * @param destinationStorageAccount Azure storage account for the destination container
+   * @param sourceContainer Azure storage container containing the blobs to be copied
+   * @param destinationContainer Azure storage container that will receive the copied blobs
+   * @return BlobCopyResult containing the status of each copy operation
+   */
+  public BlobCopierResult copyBlobs(
+      ControlledAzureStorageResource sourceStorageAccount,
+      ControlledAzureStorageResource destinationStorageAccount,
+      ControlledAzureStorageContainerResource sourceContainer,
+      ControlledAzureStorageContainerResource destinationContainer) {
+    var sourceBlobContainerClient =
+        storageAccessService.buildBlobContainerClient(sourceContainer, sourceStorageAccount);
+    var destinationBlobCOntainerClient =
+        storageAccessService.buildBlobContainerClient(
+            destinationContainer, destinationStorageAccount);
+
+    // filter out any zero-length blobs, these are not copyable
+    var blobItems =
```

Comment:
This seems like an odd semantic. I would have expected you to create an empty blob on the target in this case. Isn't an empty file significant to some systems?

Reply:
Directories are presented as zero length blobs in azure, I'll update the comment to that effect.

Reply:
How do directories get created on the destination? If I have an empty directory in my source, it will disappear in my destination?

Reply:
Wow, OK.